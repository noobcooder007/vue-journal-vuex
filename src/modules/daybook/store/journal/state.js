export default () => ({
    isLoading: true,
    entries: [
        {
            id: new Date().getTime(),
            date: new Date().toDateString(),
            text: 'Fugiat magna irure ex nostrud. Dolore elit pariatur est ea. Adipisicing fugiat fugiat aliquip dolore adipisicing sint laboris qui anim ut. Aute enim ullamco ex minim velit officia tempor ex. Commodo dolore elit elit occaecat. Dolore quis sunt cupidatat labore.',
            picture: null
        },
        {
            id: new Date().getTime() + 1000,
            date: new Date().toDateString(),
            text: 'Aliqua velit ut tempor ut veniam eiusmod eiusmod ea in aliquip. Dolore Lorem exercitation tempor amet sit. Culpa pariatur anim proident est quis esse magna cillum non veniam ullamco exercitation veniam laborum. Occaecat excepteur incididunt Lorem commodo.',
            picture: null
        },
        {
            id: new Date().getTime() + 1000,
            date: new Date().toDateString(),
            text: 'Pariatur reprehenderit ad reprehenderit sit et amet culpa officia mollit aute ad aliquip quis eu. Elit adipisicing reprehenderit amet sunt tempor. Nulla anim Lorem ipsum elit aliquip ex dolor fugiat amet veniam sunt.',
            picture: null
        }
    ]
})