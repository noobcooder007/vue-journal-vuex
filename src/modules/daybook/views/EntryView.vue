<template>
	<template v-if="entry">
		<div class="entry-title d-flex justify-content-between p-2">
			<div>
				<span class="text-success fs-3 fw-bold">{{ day }}</span>
				<span class="mx-1 fs-3">{{ month }}</span>
				<span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
			</div>
			<div>
				<button class="btn btn-danger mx-2" @click="onDeleteEntry" v-if="entry.id">
					Borrar
					<i class="fa fa-trash-alt"></i>
				</button>
				<input type="file" @change="onSelectedImage" ref="imageSelector" v-show="false" accept="image/jpeg, image/png">
				<button class="btn btn-primary" @click="onSelectImage">
					Subir foto
					<i class="fa fa-upload"></i>
				</button>
			</div>
		</div>
		<hr />
		<div class="d-flex flex-column px-3 h-75">
			<textarea placeholder="¿Qué sucedió hoy?" v-model="entry.text"></textarea>
		</div>
		<!-- <img
			src="http://games4all.mx/wp-content/uploads/2021/06/girlfriend.jpg"
			alt="entry-picture"
			class="img-thumbnail"
		/> -->
		<img
			v-if="entry.picture && !localImage"
			:src="entry.picture"
			alt="entry-picture"
			class="img-thumbnail"
		/>
		<img
			v-if="localImage"
			:src="localImage"
			alt="entry-picture"
			class="img-thumbnail"
		/>
	</template>
	<Fab icon="fa-save" @on:click="saveEntry" />
</template>

<script>
import { defineAsyncComponent } from 'vue'
import { mapActions, mapGetters } from 'vuex'
import Swal from 'sweetalert2'
import getDayMonthYear from '../helpers/getDayMonthYear'
import uploadImage from '../helpers/uploadImage'
export default {
	name: 'EntryView',
	props: {
		id: {
			type: String,
			required: true
		}
	},
	components: {
		Fab: defineAsyncComponent(() => import("../components/Fab.vue")),
	},
	data() {
		return {
			entry: null,
			localImage: null,
			file: null
		}
	},
	methods: {
		...mapActions('journal', ['updateEntry','createEntry','deleteEntry']),
		loadEntry() {
			let entry
			if (this.id === 'new') {
				entry = {
					text: '',
					date: new Date().getTime()
				}
			} else {
				entry = this.getEntryById(this.id)
				if (!entry) {
					return this.$router.push({ name: 'no-entry' })
				}
			}
			this.entry = entry
		},
		async saveEntry() {
			new Swal({
				title: 'Espere por favor',
				allowOutsideClick: false
			})
			Swal.showLoading()
			// console.log('Guardando entrada')
			// console.log(this.entry)
			const picture = await uploadImage(this.file)
			this.entry.picture = picture
			if (this.entry.id) {
				await this.updateEntry(this.entry)
			} else {
				// console.log('Post de una nueva entrada')
				const entryId = await this.createEntry(this.entry)
				this.$router.push({ name: 'entry', params: { id: entryId } })
			}
			Swal.fire('Guardado', 'Entrada registrada con exito', 'success')
			this.file = null
			this.localImage = null
		},
		async onDeleteEntry() {
			// console.log('onDeleteEntry', this.entry)
			const { isConfirmed } = await Swal.fire({
				title: '¿Estas seguro?',
				text: 'Se borrara y lo perderas para siempre, eso es mucho tiempo',
				showDenyButton: true,
				denyButtonText: 'No, aún no',
				icon: 'question',
				confirmButtonText: 'Sin miedo al éxito'
			})
			if (isConfirmed) {
				Swal.fire({
					title: 'Espere por favor',
					allowOutsideClick: false
				})
				Swal.showLoading()
				await this.deleteEntry(this.entry.id)
				this.$router.push({name: 'no-entry'})
				Swal.fire('Eliminado', '', 'success')
			}
		},
		onSelectedImage(event) {
			const file = event.target.files[0]
			if (!file) {
				this.localImage = null
				this.file = null
				return
			}
			this.file = file
			const fr = new FileReader()
			fr.onload = () => this.localImage = fr.result
			fr.readAsDataURL(file)
		},
		onSelectImage() {
			this.$refs.imageSelector.click()
		}
	},
	computed: {
		...mapGetters('journal', ['getEntryById']),
		day() {
			const { day } = getDayMonthYear(this.entry.date)
			return day
		},
		month() {
			const { month } = getDayMonthYear(this.entry.date)
			return month
		},
		yearDay() {
			const { yearDay } = getDayMonthYear(this.entry.date)
			return yearDay
		}
	},
	created() {
		// console.log(this.$route
		this.loadEntry()
	},
	watch: {
		id(value, oldValue) {
			this.loadEntry()
			if (value !== oldValue) {
				this.localImage = null
			}
		}
	}
};
</script>

<style lang="scss" scoped>
textarea {
    font-size: 20px;
    border: none;
    height: 100%;
    &:focus {
        outline: none;
    }
}
img {
    width: 200px;
    position: fixed;
    bottom: 150px;
    right: 20px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>