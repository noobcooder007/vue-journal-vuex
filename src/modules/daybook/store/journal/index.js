import state from './state'
import * as actions from './actions'
import * as getters from './getters'
import * as mutations from './mutations'

const journalStoreModule = {
    namespaced: true,
    actions,
    getters,
    mutations,
    state
}

export default journalStoreModule