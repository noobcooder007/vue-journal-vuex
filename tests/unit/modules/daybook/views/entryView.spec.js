import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import Swal from 'sweetalert2'
import journal from '@/modules/daybook/store/journal'
import EntryView from '../../../../../src/modules/daybook/views/EntryView'
import { journalState } from '../../../mock-data/test-journal-state'

const createVuexStore = (initialState) => createStore({
    modules: {
        journal: {
            ...journal,
            state: {...initialState}
        }
    }
})

jest.mock('sweetalert2' , () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('Pruebas en el EntryView', () => {
    const store = createVuexStore(journalState)
    store.dispatch = jest.fn()
    const mockRouter = {
        push: jest.fn()
    }
    let wrapper
    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: journalState.entries[2].id
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })
    test('Debe de sacar al usuario porque el id no existe', () => {
        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'Este ID no existe en el store'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
        expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' })
    })
    test('Debe de mostrar la entrada correctamente', () => {
        // console.log(wrapper.html())
        expect(wrapper.html()).toMatchSnapshot()
        expect(mockRouter.push).not.toHaveBeenCalled()
    })
    test('Debe de borrar la entrada y salir', async () => {
        Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }))
        await wrapper.find('.btn-danger').trigger('click')
        expect(Swal.fire).toHaveBeenCalledWith({
            title: '¿Estas seguro?',
            text: 'Se borrara y lo perderas para siempre, eso es mucho tiempo',
            showDenyButton: true,
            denyButtonText: 'No, aún no',
            icon: 'question',
            confirmButtonText: 'Sin miedo al éxito'
        })
        expect(mockRouter.push).toHaveBeenCalled()
        expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', journalState.entries[2].id)
    })
})