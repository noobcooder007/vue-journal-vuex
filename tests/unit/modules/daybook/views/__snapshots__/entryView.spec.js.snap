// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView Debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">18</span><span class="mx-1 fs-3">Agosto</span><span class="mx-2 fs-4 fw-light">2021, Miércoles</span></div>
  <div><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><input type="file" accept="image/jpeg, image/png" style="display: none;"><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?"></textarea></div>
<!-- <img
			src="http://games4all.mx/wp-content/uploads/2021/06/girlfriend.jpg"
			alt="entry-picture"
			class="img-thumbnail"
		/> --><img src="https://res.cloudinary.com/ddh2mpcsq/image/upload/v1629298236/tmldlajivkgl8rsycfd1.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
